stepflow-rs/crates/stepflow-typecheck/src/error.rs

```diff
@@ -0,0 +1,256 @@
+// or implied. See the License for the specific language governing permissions and limitations under
+// the License.
+
+//! Type checking errors.
+//!
+//! These errors are specific to the type checking process. They are designed
+//! to be mapped to diagnostic messages when integrated with `stepflow-analysis`.
+
+use thiserror::Error;
+
+use crate::Type;
+
+/// Errors produced during type checking.
+#[derive(Error, Debug, Clone, PartialEq)]
+pub enum TypeError {
+    /// Type mismatch between expected and actual types.
+    #[error("expected {expected}, found {actual}: {detail}")]
+    Mismatch {
+        /// The expected type.
+        expected: Type,
+        /// The actual type that was found.
+        actual: Type,
+        /// Additional detail about the mismatch.
+        detail: String,
+    },
+
+    /// Property not found when projecting through a path.
+    #[error("property '{property}' not found in path '{path}'")]
+    PropertyNotFound {
+        /// The full path being accessed.
+        path: String,
+        /// The property that was not found.
+        property: String,
+    },
+
+    /// Array index access on non-array type.
+    #[error("cannot index into non-array type at '{path}'")]
+    NotAnArray {
+        /// The path where array access was attempted.
+        path: String,
+    },
+
+    /// Unknown step referenced in expression.
+    #[error("unknown step '{step_id}'")]
+    UnknownStep {
+        /// The step ID that was not found.
+        step_id: String,
+    },
+
+    /// Unknown variable referenced in expression.
+    #[error("unknown variable '{variable}'")]
+    UnknownVariable {
+        /// The variable name that was not found.
+        variable: String,
+    },
+
+    /// Cannot determine type statically.
+    #[error("cannot determine type: {reason}")]
+    Indeterminate {
+        /// The reason the type cannot be determined.
+        reason: String,
+    },
+
+    /// Missing required property in input.
+    #[error("missing required property '{property}'")]
+    MissingRequired {
+        /// The property that is required but missing.
+        property: String,
+    },
+
+    /// Extra property in strict object.
+    #[error("unexpected property '{property}' at '{path}', allowed: {allowed:?}")]
+    UnexpectedProperty {
+        /// The property that is not allowed.
+        property: String,
+        /// The path where the property was found.
+        path: String,
+        /// The allowed properties (if known).
+        allowed: Vec<String>,
+    },
+
+    /// Component output schema is not available.
+    #[error("component '{component}' does not provide output schema")]
+    UntypedOutput {
+        /// The component that lacks output schema.
+        component: String,
+    },
+}
+
+impl TypeError {
+    /// Create a type mismatch error.
+    pub fn mismatch(expected: Type, actual: Type, detail: impl Into<String>) -> Self {
```

Comment:
Probably easier to create directly to benefit from named parameters (for this and other type errors).